enterprise/reporting/reporting.go

```diff
@@ -361,126 +377,116 @@ func (r *DefaultReporter) mainLoop(ctx context.Context, c types.SyncerConfig) {
 	r.requestLatency = r.stats.NewTaggedStat(StatReportingHttpReqLatency, stats.TimerType, tags)
-	reportingLag := r.stats.NewTaggedStat(
-		"reporting_metrics_lag_seconds", stats.GaugeType, stats.Tags{"client": c.Label},
-	)
 
 	var lastReportedAtTime atomic.Time
 	lastReportedAtTime.Store(time.Now())
 
-	var wg sync.WaitGroup
-	wg.Add(1)
-	defer wg.Wait()
-	go func() {
-		// for monitoring reports pileups
-		defer wg.Done()
-		for {
-			lag := time.Since(lastReportedAtTime.Load())
-			reportingLag.Gauge(lag.Seconds())
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(2 * time.Minute):
-			}
-		}
-	}()
+	g, ctx := errgroup.WithContext(ctx)
 
-	for {
-		if ctx.Err() != nil {
-			r.log.Infof("stopping mainLoop for syncer %s : %s", c.Label, ctx.Err())
-			return
-		}
-		requestChan := make(chan struct{}, r.maxConcurrentRequests.Load())
-		loopStart := time.Now()
-		currentMin := time.Now().UTC().Unix() / 60
-
-		getReportsStart := time.Now()
-		reports, reportedAt, err := r.getReports(currentMin, c.ConnInfo)
-		getReportsTimer.Since(getReportsStart)
-		getReportsCount.Observe(float64(len(reports)))
-		if len(reports) == 0 {
-			if err == nil {
-				lastReportedAtTime.Store(loopStart)
-			} else {
-				r.log.Errorw("getting reports", "error", err)
-			}
-			select {
-			case <-ctx.Done():
+	g.Go(func() error {
+		return r.emitLagMetric(ctx, c, &lastReportedAtTime)
+	})
+
+	g.Go(func() error {
+		for {
+			if ctx.Err() != nil {
 				r.log.Infof("stopping mainLoop for syncer %s : %s", c.Label, ctx.Err())
-				return
-			case <-time.After(r.sleepInterval.Load()):
+				return ctx.Err()
 			}
-			continue
-		}
-
-		lastReportedAtTime.Store(time.Unix(reportedAt*60, 0))
-		getAggregatedReportsStart := time.Now()
-		metrics := r.getAggregatedReports(reports)
-		getAggregatedReportsTimer.Since(getAggregatedReportsStart)
-		getAggregatedReportsCount.Observe(float64(len(metrics)))
-
-		errGroup, errCtx := errgroup.WithContext(ctx)
-		// default to -1 to allow unlimited concurrency
-		errGroup.SetLimit(config.GetInt("Reporting.maxConcurrentRequests", -1))
-		for _, metric := range metrics {
-			if r.whActionsOnly && metric.SourceCategory != "warehouse" {
-				// if whActionsOnly is true, we only send reports for wh actions sources
-				// we silently drop all other reports
+			requestChan := make(chan struct{}, r.maxConcurrentRequests.Load())
+			loopStart := time.Now()
+			currentMin := time.Now().UTC().Unix() / 60
+
+			getReportsStart := time.Now()
+			reports, reportedAt, err := r.getReports(currentMin, c.ConnInfo)
+			getReportsTimer.Since(getReportsStart)
+			getReportsCount.Observe(float64(len(reports)))
+			if len(reports) == 0 {
+				if err == nil {
+					lastReportedAtTime.Store(loopStart)
+				} else {
+					r.log.Errorw("getting reports", "error", err)
+				}
+				select {
+				case <-ctx.Done():
+					r.log.Infof("stopping mainLoop for syncer %s : %s", c.Label, ctx.Err())
```

Comment:
I notice an inconsistent behaviour here:
1. in line 394 https://github.com/rudderlabs/rudder-server/pull/4457/files#diff-5639a722fc70b6a04ec6daf63094d6bb127454dd8587e1b3466785305e515c3eR392 , when context is Err, you return an error
2. here `<-ctx.Done()` you return nil. 

ctx.Err() and ctx.Done() are very similar in how they work, except done is blocking until context gets cancels and .Err() returns a not nil error when context gets cancel.

Also to avoid spamming, I would avoid logging an error and returning it, as well. It would be best if you decided how your component should behave and be consistent with it.